File: Linked-List/LinkedList.java
class Node<T>
{
    public T data;
    public Node<T> next;

    public Node(T data, Node<T> p)
    {
        this.data = data;
        this.next = p;
    }
}

class LinkedList<T>
{

    protected Node<T> head = null;
    private int size = 0;

    public void add_start(T data)
    {
        Node<T> ptr = new Node<>(data,head);
        head = ptr;
        size++;
    }

    public void add_end(T data)
    {
        /*
        Creating a temp node and then traversing to the end node and adding the new node
        */
        if (head!=null)
        {
            Node<T> temp = head;
            while(temp.next!=null)
                temp = temp.next;
            temp.next = new Node<>(data,null);
            size++;
        }
        else
            // If there is no element then add at start
            add_start(data);
    }

    public void add(int index, T data)
    {
        if (index >size || index < 0)
        {
            System.out.println("Index out of range");
            System.exit(0);
        }
        // Traversing to the index and adding a new node and poiting the new node to the next elemnt
        if (index == 0 )
        {
            add_start(data);
            return;
        }
        Node <T>temp = head;
        for (int i=0;i<size;i++)
        {
            if (i == index-1)
            {
                temp.next = new Node<>(data,temp.next);
                break;
            }
            temp = temp.next;
        }
        size++;
    }

    public void remove(int index)
    {
        if (index >size-1 || index < 0)
        {
            System.out.println("Index out of range");
            System.exit(0);
        }
        // Traversing to the index before the entered index and poiting it to the next element skipping the current index
        Node<T> temp = head;
        if (index == 0)
            head = head.next;
        else
        {
            for (int i=0; i<size-1; i++)
            {
                if (i == index-1)
                {
                    temp.next = temp.next.next;
                    break;
                }
                temp = temp.next;
            }  
        }

        size--;
    }

    public int length()
    {
        return size;
    }

    public void display()
    {
        if (head!=null)
        {
            Node<T> temp = head;
            while(temp.next!=null)
            {
                System.out.print(temp.data + " -> ");
                temp = temp.next;
            }
            System.out.println(temp.data + " -> *");
        }
        
    }

    public T getdata(int index)
    {
        if (index >size-1 || index < 0)
        {
            System.out.println("Index out of range");
            System.exit(0);
        }

        Node<T> temp = head;

        for (int i=0; i<size;i++)
        {
            if (i == index)
                return temp.data;
            temp = temp.next;
        }

        T t=null;
        return t;
    }
}


class Iterator<T>
{

    private Node<T> node;
    private Node<T> head;
    private int size;
    private int current = 0;
    Iterator(LinkedList<T> n)
    {
        node = n.head;
        head = n.head;
        size = n.length();
    }

    boolean has_next()
    {
        if (current<size)
        {
            return true;
        }
        else
        {
            current = 0;
            node = head;
            return false;
        }
        
    }

    T data()
    {
        T temp = node.data;
        node = node.next;
        current++;
        return temp;
    }
}

class LinkedListTest
{
    public static void main(String[] args)
    {
        LinkedList<Double> l = new LinkedList<>();
        l.display();
        l.add_end(0.1);
        l.add_start(10.1);
        l.add_end(80.1);
        l.add_start(20.1);
        l.add_start(30.1);
        l.add_start(40.1);
        l.add_end(50.1);
        l.add_start(130.1);
        l.display();
        l.add(5,100.1);
        l.display();
        l.remove(3);
        l.display();
        l.remove(7);
        l.display();
        System.out.println(l.getdata(1));
        Iterator <Double> Iter = new Iterator<>(l);

        while(Iter.has_next())
        {
            System.out.println(Iter.data());
        }
    }
}

File: Linked-List/linkedlist.py
#!/usr/bin/python3
class Node:
    def __init__(self,data=None,next=None):
        self.data = data
        self.next = next
        

class LinkedList:
    def __init__(self):
        self.head = None
        self.size = 0
    
    def add_start(self,data):
        ptr = Node(data,self.head)
        self.head = ptr
        self.size += 1
    
    def add_end(self,data):
        if not self.head:
            # If there is no element then add at start
            self.add_start(data)
            return
        # Creating a temp node and then traversing to the end node and adding the new node
        temp = self.head
        while(temp.next):
            temp = temp.next
        temp.next = Node(data,None)
        self.size += 1 

    def add(self,index,data):
        assert not (index > self.size or index < 0), "Index out of range"
        
        # Traversing to the index and adding a new node and poiting the new node to the next elemnt
        if index == 0:
            self.add_start(data)
            return
        temp = self.head
        for i in range(self.size):
            if i == index-1:
                temp.next = Node(data,temp.next)
                break
            temp = temp.next
        
        self.size += 1
  
    def remove(self,index):
        assert not (index > self.size-1 or index < 0), "Index out of range"
        # Traversing to the index before the entered index and poiting it to the next element skipping the current index
        temp = self.head
        if index == 0:
            self.head = self.head.next
        else:
            for i in range(self.size-1):
                if i == index-1:
                    temp.next = temp.next.next
                    break
                temp = temp.next
            
        self.size -= 1
  
    def length(self):
        return self.size

    def display(self):
        if self.head:
            temp = self.head
            while(temp.next):
                print(temp.data,"->",end=" ")
                temp = temp.next
            print(temp.data,"->",end=" ")
            print("*")
        
    def getdata(self,index):
        assert not ( index > self.size-1 or index < 0), "Index out of range"
        
        temp = self.head
        for i in range(self.size):
            if i == index:
                return temp.data
            temp = temp.next
        
class Iterator:
    def __init__(self,n):
        self.node = n.head
        self.head = n.head
        self.size = n.length()
        self.current = 0
        
    def has_next(self):
        if self.current < self.size:
            return True
        else:
            self.current = 0
            self.node = self.head
            return False
        
    def data(self):
        temp = self.node.data
        self.node = self.node.next
        self.current += 1
        return temp
    
l = LinkedList()
l.display()
l.add_end(0)
l.add_start(10)
l.add_end(80)
l.add_start(20)
l.add_start(30)
l.add_start(40)
l.add_end(50)
l.add_start(130)
l.display()
l.add(5,100)
l.display()
l.remove(3)
l.display()
l.remove(7)
l.display()
print(l.getdata(1))
Iter = Iterator(l)

while Iter.has_next():
    print(Iter.data())